### PR TITLE
Non-blocking subscribe (#on_each_message)

### LINKED ISSUE #29: Non-blocking subscribe (#on_each_message)
For Twingly::AMQP::Subscription

E.g. `#on_each_message`, similar to the `#on_exception` we have

### PR DESCRIPTION
**Pull request by roback**
I didn't want to duplicate all tests from `#each_message`, so I just copied the one I felt was most important (the one that makes sure it can receive a message from rabbitmq).

close #29

### CODE REVIEW

**Comment by walro** [comment]
have you considered it to be the one and same method and blocking is provided as a parameter (possibly to initialize)? Similar to how Bunny works that is.

File: spec/integration/twingly/amqp/subscription_spec.rb [hunk 1]
```diff
@@ -147,4 +151,32 @@
...
+    it "should receive the message published on the exchange" do
+      exchange.publish(payload_json, routing_key: routing_key)
+      exchange.wait_for_confirms
+
+      received_url = nil
+      subject.on_each_message do |message|
+        received_url = message.payload[:url]
+      end
+
+      wait_for { !received_url.nil? }
```
Comment (walro) [inline spec/integration/twingly/amqp/subscription_spec.rb]: I think if the ordering is changed we can do without the `wait_for`.

```diff
diff --git a/spec/integration/twingly/amqp/subscription_spec.rb b/spec/integration/twingly/amqp/subscription_spec.rb
index 09277b3..9284c67 100644
--- a/spec/integration/twingly/amqp/subscription_spec.rb
+++ b/spec/integration/twingly/amqp/subscription_spec.rb
@@ -10,10 +10,6 @@ describe Twingly::AMQP::Subscription do
     end
   end
 
-  def wait_for
-    sleep 0.1 until yield
-  end
-
   include_context "amqp queue"
 
   let(:payload_url)    { "http://www.test.se" }
@@ -166,15 +162,14 @@ describe Twingly::AMQP::Subscription do
     end
 
     it "should receive the message published on the exchange" do
-      exchange.publish(payload_json, routing_key: routing_key)
-      exchange.wait_for_confirms
-
       received_url = nil
+
       subject.on_each_message do |message|
         received_url = message.payload[:url]
       end
 
-      wait_for { !received_url.nil? }
+      exchange.publish(payload_json, routing_key: routing_key)
+      exchange.wait_for_confirms
 
       expect(received_url).to eq(payload_url)
     end
```

**Comment by roback** [comment]
> have you considered it to be the one and same method and blocking is provided as a parameter (possibly to initialize)? Similar to how Bunny works that is.

I have not. That would be more clear to the user of this gem what is happening. `each_message(blocking: true)` is more intention revealing than `on_each_message`.

File: spec/integration/twingly/amqp/subscription_spec.rb [hunk 2]
```diff
@@ -146,5 +146,32 @@
...
+
+      it "should receive the message published on the exchange" do
+        received_url = nil
+        subject.each_message(blocking: false) do |message|
+          received_url = message.payload[:url]
+        end
+
+        exchange.publish(payload_json, routing_key: routing_key)
+        exchange.wait_for_confirms
+
```
Comment (walro) [inline spec/integration/twingly/amqp/subscription_spec.rb]: maybe test publishing another message to verify that it works for more than one time

File: lib/twingly/amqp/subscription.rb [hunk 3]
```diff
@@ -24,24 +24,13 @@ def initialize(queue_name:, exchange_topic: nil, routing_key: nil, consumer_thre
...
-          message = Message.new(
-            delivery_info: delivery_info,
-            metadata:      metadata,
-            payload:       payload,
-            channel:       @channel,
-          )
-
-          block.call(message)
-        end
+        return unless blocking
```
Comment (walro) [inline lib/twingly/amqp/subscription.rb]: WDYT of structuring like this instead:

```ruby
if (blocking)
  sleep 0.01 until cancel?
  consumer.cancel
end
```

Comment (roback) [inline lib/twingly/amqp/subscription.rb]: Sure, makes it more obvious that those lines belong together.

I don't think RuboCop would approve of the parenthesis in the `if` statement though 😛

Comment (walro) [inline lib/twingly/amqp/subscription.rb]: > I don't think RuboCop would approve of the parenthesis

I removed those a long time ago :P (directly after posting). GitHub is not that great at updating.

Comment (roback) [inline lib/twingly/amqp/subscription.rb]: Haha 😄

**Pull request merged**